### PR TITLE
refactor: use patch package instead of hardcoded operations

### PR DESCRIPTION
**Pull request by alicefr**
### What this PR does
Before this PR:
In the code we have in multiple places hardcoded patches with  [`GeneratePatchBytes`](https://github.com/kubevirt/kubevirt/blob/main/pkg/controller/controller.go#L311) or  [`GenerateTestReplacePatch`](https://github.com/kubevirt/kubevirt/blob/main/pkg/apimachinery/patch/patch.go#L54)

After this PR:
The code will be much consistent by using the new API all over the places and we eliminate hardcoded patch operations

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/watch/migration.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration.go:250]: `Printf`

File: pkg/virt-controller/watch/migration.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration.go:251]: Possibly rename / create a new package called `patchset`? IMO `patchset.New()` reads better than `patch.NewPatchSet()`.

File: pkg/virt-controller/watch/migration.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration.go:266]: nit: newline

File: pkg/virt-controller/watch/migration.go
Comment (0xFelix) [reply pkg/virt-controller/watch/migration.go:250]: ... and some more down below

File: pkg/virt-controller/watch/migration_test.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration_test.go:442]:
```suggestion
		var migrationConfiguration *virtv1.MigrationConfiguration
		if len(customConfigs) > 0 {
			migrationConfiguration = customConfigs[0]
		} else {
			migrationConfiguration = controller.clusterConfig.GetMigrationConfiguration()
		}
		Expect(migrationConfiguration).ToNot(BeNil())
```

File: pkg/virt-controller/watch/migration_test.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration_test.go:1683]: This looks like a huge improvement to me! :+1:

File: pkg/virt-controller/watch/migration_test.go
Comment (0xFelix) [inline pkg/virt-controller/watch/migration_test.go:255]: Looks like another improvement to me, can we replace all occurences with this?

**Review by 0xFelix — COMMENTED** [review]
Very good idea! I like it.

**Comment by alicefr** [comment]
@0xFelix sorry this isn't ready for review. I wanted to check if the migration functional tests were fine with this change

File: pkg/virt-controller/watch/migration.go
Comment (alicefr) [reply pkg/virt-controller/watch/migration.go:250]: Yes, this is a leftover. You were too quick with the review. I should have written more explicitly that this is WIP

File: pkg/virt-controller/watch/migration.go
Comment (alicefr) [reply pkg/virt-controller/watch/migration.go:250]: Actually, I'd like to replace everything with the patchset interface and remove completely also the `GenerateTestReplacePatch` and the types. About the naming, when I have refactor the code, it should be easy to rename the patch.go and the package to patchset

File: pkg/virt-controller/watch/vmi.go
Comment (prnaraya) [inline pkg/virt-controller/watch/vmi.go:466]: super tiny nit: in line 445 you refer to this as "patch"  but here you use patchBytes for what appears to be the same thing. It might be good to make this consistent across both?

**Review by prnaraya — COMMENTED** [review]
Thanks for the PR! I like where this refactoring is going. Just left a couple small comments inline.

File: pkg/virt-api/rest/subresource_test.go
Comment (alicefr) [reply pkg/virt-api/rest/subresource_test.go:958]: Oh yes, I missed it. Thanks!

**Review by vasiliy-ul — COMMENTED** [review]
Great work with this PR :+1:

File: pkg/virt-controller/watch/migration.go
Comment (alicefr) [reply pkg/virt-controller/watch/migration.go:264]: @vasiliy-ul it's true but the operation are distinct, that's why I'm a bit reluctant to put them together

Comment (vasiliy-ul) [reply pkg/virt-controller/watch/migration.go:264]: Just to make it clear: what I am proposing here is to extend the patchSet struct with `TestAndRepalce` call. Still keeping individual methods for `Test` and `Replace`. That way, code snippets like the one here can be simplified.

But I do not insist. Up to you.

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go]: Done

File: pkg/virt-api/rest/subresource.go
Comment (vasiliy-ul) [reply pkg/virt-api/rest/subresource.go:1358]: Maybe smth like:

```go
func AddOrReplace(path string, value interface{}, condition bool) {
	if condition {
		patches.Add(path, value)
	} else {
		patches.Replace(path, value)
	}
}
...
// actual function call will look like:
AddOrReplace("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest, vm.Status.MemoryDumpRequest == nil)
```

Comment (alicefr) [reply pkg/virt-api/rest/subresource.go:1358]: Ok, sounds good, I'll add it

File: pkg/virt-api/rest/subresource.go
Comment (alicefr) [reply pkg/virt-api/rest/subresource.go]: I renamed the function `NewSet` to `New` but I l left `patch` as it is the name of the file and there are other functions declared there. If I rename the package patchset, then also the other functions are from `patches`. Anyway, this was my reasoning behind it :)

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:104]: Done!

File: pkg/virt-api/rest/subresource.go
Comment (alicefr) [reply pkg/virt-api/rest/subresource.go]: There are still cases  where there is also Add or Test  + Replace. Do I need to add this too?I seems a bit too much but I can .

File: pkg/virt-api/rest/subresource_test.go
Comment (alicefr) [reply pkg/virt-api/rest/subresource_test.go]: No, you are right , I simply removed it

File: pkg/virt-api/rest/subresource_test.go
Comment (0xFelix) [inline pkg/virt-api/rest/subresource_test.go:1442]: Makes sense to add `patches.TestAndAdd` since we also have `patches.TestAndReplace`?

Comment (alicefr) [reply pkg/virt-api/rest/subresource_test.go:1442]: Not sure. We should also then add `AddOrTestAndReplace`. It because a bit lengthy imo

Comment (0xFelix) [reply pkg/virt-api/rest/subresource_test.go:1442]: `patches.TestAndAdd` sounds like a low hanging fruit to me.

**Review by 0xFelix — COMMENTED** [review]
Thanks, it looks great to me! But since it is such a large PR I would feel more comfortable if we could get a second reviewer for LGTM. Would that be ok for you?

**Comment by alicefr** [comment]
> Thanks, it looks great to me! But since it is such a large PR I would feel more comfortable if we could get a second reviewer for LGTM. Would that be ok for you?

Sure!
@vasiliy-ul since you already reviewed this, do you mind having a second look?

File: pkg/virt-controller/watch/vm.go
Comment (vasiliy-ul) [inline pkg/virt-controller/watch/vm.go:2209]: nit: does it make sense to keep this function? Maybe better to call `patch.PatchTestReplaceFinalizers` directly where needed?

File: pkg/virt-operator/resource/apply/core_test.go
Comment (vasiliy-ul) [inline pkg/virt-operator/resource/apply/core_test.go:401]: nit: Maybe better to call `strings.Contains(...)` directly?

**Review by vasiliy-ul — APPROVED** [review]
The code looks good overall. Added a couple of minor remarks. All optional. Feel free to ignore if you disagree.

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:150]: TBH, I would keep the And because we have a function with an Or. This makes it clearer imo.

File: pkg/virt-controller/watch/vm.go
Comment (alicefr) [reply pkg/virt-controller/watch/vm.go]: Done, thanks!

**Review by vasiliy-ul — APPROVED** [review]

**Comment by alicefr** [comment]
Rebased

**Comment by 0xFelix** [comment]
Readding LGTM again

/lgtm

File: pkg/virt-api/rest/subresource.go
Comment (0xFelix) [inline pkg/virt-api/rest/subresource.go:226]:
```suggestion
			return nil, fmt.Errorf("unable to complete request: stop/start already underway")
```

nit

**Comment by alicefr** [comment]
Rebased and fixing the patch arguments

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Comment by alicefr** [comment]
Rebased to include https://github.com/kubevirt/kubevirt/pull/11517

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Comment by alicefr** [comment]
@vasiliy-ul would you mind to take another look and possibly approve? There hasn't be any major changes

**Review by vasiliy-ul — COMMENTED** [review]
/approve

**Comment by vasiliy-ul** [comment]
@alicefr, `pull-kubevirt-e2e-kind-sriov` has failed several times in a row with the same set of tests. This probably needs a closer look. Adding a hold. Feel free to remove if it is not related to this PR.

/hold

**Comment by alicefr** [comment]
Rebased

**Comment by alicefr** [comment]
> @alicefr, `pull-kubevirt-e2e-kind-sriov` has failed several times in a row with the same set of tests. This probably needs a closer look. Adding a hold. Feel free to remove if it is not related to this PR.
> 
> /hold

Mmmh yes there is a panic in virt-controller. Looking into it

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [inline pkg/apimachinery/patch/patch.go:40]: Interfaces are usually expressing behavior and not a set of methods for an object.
In Go, it is consider a good-practice to define them in the place they are used, and not in the package that defines their concrete.

Sometime, the solution is to have a dedicated package for interfaces, but IMO it should be used only in rare cases. The Go std have some examples of this.
But even there, there are multiple interfaces defined, not all-in-one.

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [inline pkg/apimachinery/patch/patch.go:56]: It should return a concrete type, not an interface. The caller can use it through an interface if it wants to.

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [inline pkg/apimachinery/patch/patch.go:93]: This is odd.
Why do we need a condition and not a direct usage?

Also, in most cases, `Replace` is not really needed, `Add` is enough and will do both (i.e. add if missing or override is already exists).

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [inline pkg/apimachinery/patch/patch.go:88]: Similar to `AddOrReplace`, if one does a `test`, I find it hard to see a use case that needs`replace`. An `add` will work better in most cases.

But I would like to propose that this object will not have all these combinations, it should just support the basic commands.

A factory can be added to build a collections of operations, with a build pattern. Then, whatever combination is required, it can be built and avoid having such a matrix of methods.

Example:
```go
patchSet := patch.New(
  patch.WithAdd(...)
  patch.WithTest(...)
)

vm, err := kubevirt.Client().VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchSet.Payload(), &metav1.PatchOptions{}
```

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:40]: I am also missing the unit tests for this new object, expressing how it is expected to be used and showing basic scenarios.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:56]: Ok, I can make it a type not an interface

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: @EdDev see for example this discussion: https://github.com/kubevirt/kubevirt/pull/11211#discussion_r1502554835

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:93]: Yes, I know in many places `replace` is used. My claim is that it is not needed.
There are also many places that `add`/`replace` is combined with `test` without a good reason.

File: pkg/apimachinery/patch/patch.go
Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:88]: The example you linked is good. The code there is wrong and redundant.

```go
if vm.Status.MemoryDumpRequest == nil {
	patches.Add("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
} else {
	patches.Replace("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
```

Is equivalent to a simple:
```go
patches.Add("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
```

`add` overrides the existence, same as `replace`.
The only difference between the two is that `replace` will blow up if the path was not there before.

**Comment by alicefr** [comment]
Convert to draft until we clarify https://github.com/kubevirt/kubevirt/pull/11211#discussion_r1538650935

File: pkg/apimachinery/patch/patch.go
Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: Are we sure? Checking the [documentation](https://jsonpatch.com/#operations), it says `Replaces a value. Equivalent to a “remove” followed by an “add”.`

Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: I thought `add` should fail if the value already exists and not replacing it. I'll give a try with a small example

Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:88]: > Are we sure? Checking the [documentation](https://jsonpatch.com/#operations), it says `Replaces a value. Equivalent to a “remove” followed by an “add”.`

Yes, `remove` will fail if the given path is not there and that is essentially the main different between the two.

Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: Ok, thanks! Make sense and we can further simplify the code. I just tried to stick to what we had in the previous version of the code.

I would also need to heavily change the unit tests as basically the check the same sequence of patches we have in the flow.

Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:88]: IMO it will help if you first create the initial object and use it to cover all its API. Then, in follow up PRs replace to use it.

Trying to do everything in one PR is too challenging and it will be hard to merge without rebasing all the time.

Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: Not sure, I understand. Do you meant just to add the new type in the code with the unit tests and then in a follow-up PR replacing the existing operations?

Comment (alicefr) [reply pkg/apimachinery/patch/patch.go:88]: @EdDev @0xFelix something like: https://github.com/kubevirt/kubevirt/pull/11599 ? We could initially replace all the `GeneratePatchPayload` and then in a following PR all the hardcoded operations as I did in this PR. WDYT?

Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:88]: My suggestion was to create the new object with all its methods and its unit tests.
Continue in other commits to replace places in the codebase to use it... at least one time each method.
This should give the reviewers the chance to see how it is used and convince them it makes sense.

Then, in follow up PRs to start replacing places in the codebase one by one (or as a group).

**Comment by alicefr** [comment]
Closed in favor of https://github.com/kubevirt/kubevirt/pull/11599

**Pull request closed without merging**
